### PR TITLE
Implementation of instantiation evaluation state

### PR DESCRIPTION
**Pull request by ajreynol**
This implements the core notification methods for term evaluation and tracking whether quantifiers are still active.

### CODE REVIEW

File: src/theory/quantifiers/ieval/state.h [hunk 1]
```diff
@@ -87,6 +87,15 @@ class State : protected EnvObj
   bool isSome(TNode n) const;
   /** Invoke the rewriter for term n */
   Node doRewrite(Node n) const;
+  /** Is quantifier active? */
+  bool isQuantActive(TNode q) const;
+  /** Set quantified formula active / inactive */
```
Comment (HanielB) [inline src/theory/quantifiers/ieval/state.h:92]:
```suggestion
  /** Set quantified formula inactive */
```

File: src/theory/quantifiers/ieval/state.cpp [hunk 2]
```diff
@@ -133,6 +313,171 @@ const PatTermInfo& State::getPatTermInfo(TNode p) const
   return it->second;
 }
 
+void State::notifyPatternEqGround(TNode p, TNode g)
+{
+  Assert(!g.isNull());
+  Assert(!expr::hasBoundVar(g));
+  Assert(d_tec->evaluateBase(*this, g) == g);
```
Comment (HanielB) [inline src/theory/quantifiers/ieval/state.cpp:320]: IIUC this is not capturing to when `g` is `some`? My understanding from the docs is that it cannot be `some`, has to either ground or `none`.

**Review by HanielB — APPROVED** [review]
LGTM, just some minor comments

File: src/theory/quantifiers/ieval/state.cpp (hunk 2 above)
Comment (ajreynol) [reply src/theory/quantifiers/ieval/state.cpp:320]: It can be some, the doc was incomplete, I've updated.

**Pull request merged**